### PR TITLE
feat: Add chunk associations task for memify pipeline

### LINKED ISSUE #1333: Add a task that creates common association from a group of chunks
# Chunk Associations Task

***NOTE:** This issue is part of Contribute-to-Win. Please comment first to get assigned. Read the details [here](https://github.com/orgs/topoteretes/discussions/1361)*

A task for creating semantic associations between document chunks and organizing them into common node sets based on their similarity.

## Overview

The Chunk Associations Task analyzes document chunks for semantic similarity and creates weighted association edges between them. It creates weighted edgees between highly associated chunks with value Association and uses LLM classifier that has a prompt behind that should have an output saying if we should create an association or not.

Example

Book 1: Chunk talks about river dolphins
Book 2: Chapter about river dolphins

Book 1 - Chunk 1 -> Association weight to Chapter about river dolphin

Input

The task should receive a batch of N chunks retrieved from an existing graph

Output

Datapoints with weighted edges that can be stored in the graph. They can belong to special Associaton nodeset or we can update and delete existing data.

### PR DESCRIPTION
**Pull request by deidaraiorek**
## Description

This PR adds a new enrichment task (`create_chunk_associations`) that analyzes document chunks for semantic similarity using LLM-based comparison and creates weighted "associated_with" edges between them in the knowledge graph.

Will close #1333

The task uses vector search for candidate selection to find semantically related chunks, then uses an LLM to verify and score the similarity.

## Acceptance Criteria

* Configurable similarity threshold (default 0.7)
* Configurable min chunk length and top-k candidates
* Creates weighted edges with association type and reasoning metadata
* Integrates as enrichment task in the memify pipeline

### How to verify

Run the example script:
```bash
python examples/python/chunk_associations_example.py
```

### Example output

```
Adding text chunks to cognee:

1. River dolphins are freshwater mammals found in South America and Asia.
2. Scientists study dolphin behavior in Amazon rivers to understand their communication.
3. Python is a high-level programming language widely used for data science.
4. Dolphins use echolocation to navigate and hunt in murky river waters.
5. Machine learning models require large datasets for training.
6. The Amazon river dolphin, also known as boto, is pink in color.

Text added successfully.

Running cognify to create knowledge graph...
Cognify process complete.

Running memify to create chunk associations...
Memify process complete. Querying for chunk associations...

Found 6 chunk associations:

1. Similarity: 0.85
   Type: topical
   Chunk 1: River dolphins are freshwater mammals found in South America and Asia.
   Chunk 2: Scientists study dolphin behavior in Amazon rivers to understand their communica
   Reasoning: Chunks are semantically related as they both discuss dolphins and their habitats, with chunk 2 providing more specific information on research related to river dolphins.

2. Similarity: 0.85
   Type: topical
   Chunk 1: River dolphins are freshwater mammals found in South America and Asia.
   Chunk 2: The Amazon river dolphin, also known as boto, is pink in color.
   Reasoning: Chunks 1 and 2 are clearly related as they discuss the same species, Amazon river dolphins/boto.

3. Similarity: 0.85
   Type: topical
   Chunk 1: River dolphins are freshwater mammals found in South America and Asia.
   Chunk 2: Dolphins use echolocation to navigate and hunt in murky river waters.
   Reasoning: Both chunks discuss river dolphins and their behavior in freshwater environments.

4. Similarity: 0.70
   Type: topical
   Chunk 1: Python is a high-level programming language widely used for data science.
   Chunk 2: Machine learning models require large datasets for training.
   Reasoning: The chunks are related as they involve the broader topic of programming and data science.

5. Similarity: 0.85
   Type: topical
   Chunk 1: Dolphins use echolocation to navigate and hunt in murky river waters.
   Chunk 2: Scientists study dolphin behavior in Amazon rivers to understand their communica
   Reasoning: Both chunks discuss dolphins, with Chunk 1 covering navigation/hunting and Chunk 2 highlighting scientific study.

6. Similarity: 0.80
   Type: topical
   Chunk 1: The Amazon river dolphin, also known as boto, is pink in color.
   Chunk 2: Scientists study dolphin behavior in Amazon rivers to understand their communica
   Reasoning: Both mention Amazon river dolphins, with Chunk 1 describing color and Chunk 2 examining behavior.
```

### Tests

Run tests:
```bash
pytest cognee/tests/test_chunk_associations.py -v
```

All 6 tests pass:
- `test_chunk_associations_creates_edges_between_similar_chunks` - verifies edges are created
- `test_chunk_associations_respects_similarity_threshold` - verifies threshold filtering
- `test_chunk_associations_includes_metadata` - verifies edge properties
- `test_chunk_associations_with_cypher_query` - verifies Cypher search works
- `test_chunk_associations_handles_single_chunk` - verifies no edges with 1 chunk
- `test_chunk_associations_configurable_parameters` - verifies min_chunk_length filtering
Screenshot 2026-02-06 at 3 32 33 PM

## Type of Change
- [x] New feature (non-breaking change that adds functionality)

## Pre-submission Checklist
- [x] **I have tested my changes thoroughly before submitting this PR**
- [x] **This PR contains minimal changes necessary to address the issue/feature**
- [x] My code follows the project's coding standards and style guidelines
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] All new and existing tests pass
- [x] I have searched existing PRs to ensure this change hasn't been submitted already
- [x] My commits have clear and descriptive messages

## DCO Affirmation
I affirm that all code in every commit of this pull request conforms to the terms of the Topoteretes Developer Certificate of Origin.

## Summary by CodeRabbit

* **New Features**
  * Added semantic chunk association: automatic detection and linking of related text chunks using vector + LLM assessments; outputs include boolean similarity, numeric score, association type, and short rationale. Configurable thresholds, min-length, candidate limits, and persistence to the knowledge graph.

* **Tests**
  * Added comprehensive tests for association creation, weight normalization, threshold behavior, metadata presence, queryability, and no-edge cases.

* **Chores**
  * Added an end-to-end example script demonstrating the chunk association workflow.

### CODE REVIEW

File: cognee/infrastructure/llm/prompts/chunk_association_system.txt [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+2. Numeric similarity score from 0.0 to 1.0
+3. Type of association (topical, causal, temporal, elaboration, contextual, or null)
+4. Brief reasoning (1-2 sentences max)
+
+Similarity score ranges:
+- 0.0-0.3: Unrelated or minimal connection
+- 0.4-0.6: Loosely related
+- 0.7-0.8: Clearly related topics
+- 0.9-0.95: Very closely related
+- 0.96-1.0: Near-duplicate or identical content
```
[the remark this answers is not in the document]
Comment (deidaraiorek) [reply cognee/infrastructure/llm/prompts/chunk_association_system.txt]: Addressed
